guacamole-common/src/main/java/org/apache/guacamole/protocol/ConfiguredGuacamoleSocket.java

```diff
@@ -83,14 +111,24 @@ private GuacamoleInstruction expect(GuacamoleReader reader, String opcode)
         if (instruction == null)
             throw new GuacamoleServerException("End of stream while waiting for \"" + opcode + "\".");
 
+        // Handle server control commands
+        if ("disconnect".equals(instruction.getOpcode()))
+            throw new GuacamoleServerException("Server disconnected while waiting for \"" + opcode + "\".");
+        if ("error".equals(instruction.getOpcode())) {
+            GuacamoleServerErrorCommandException e = parseServerErrorCommandArgs(instruction.getArgs());
+            if (e == null)
+                throw new GuacamoleServerException("Invalid command received from server: " + instruction);
```

Comment:
Might be better to let it be the responsibility of the function parsing the `error` instruction to decide whether the instruction is valid and throw an appropriate exception if it isn't. This would simplify the logic here and more cleanly separate concerns.

Reply:
Fair enough, incorporated the feedback. After a couple of iterations it does look better, have a look.